torch/csrc/autograd/profiler.cpp

```diff
@@ -13,135 +13,260 @@
+// profiler (i.e. enter profiling context manager/call enableConfig) and
+// uniquely identifies a profiling run.
+//
+// We automatically propagate ThreadLocalDebugInfo into async tasks,
+// as well as across JIT continuations and autograd thread, so all
+// the operations that happen between profiling start and end
+// (not necessarily within the same thread) are recorded.
+// Unless the profiling slot is overwritten as in the case of nested
+// profiling ranges (in this case events for the subrange are handled
+// by the nested profiler)
+//
+// When we exit a profiling range (either by exiting profiling context
+// manager or by calling disableProfiler), we remove the previously set
+// profiling entry for the given thread local mapping, and consolidate
+// events in the profiling result
+//
+//
+// ThreadLocalState:
+//
+// ThreadLocalState takes a 'snapshot' of thread local variables
+// using provided getters. It is used together with ThreadLocalStateGuard
+// to transfer the snapshot across thread boundary and set the thread local
+// values as in the parent task.
+//
+// Profiler uses ThreadLocalState to propagate profiler's thread local state.
+// Whenever we try to set the current thread local profiler config,
+// we update the corresponding thread local variable and check whether
+// we need to push or pop a pair of thread local callbacks.
+// Thus making sure that profiling callbacks are set only when needed.
+//
+//
+// RecordFunction and observers
+//
+// Profiler uses observers mechanism to add a pair of thread local conditional
+// callbacks that are executed on a fixed number of predetermined ranges,
+// including:
+//  - c10/ATen ops
+//  - TorchScript functions/methods
+//  - user defined named ranges (see `record_function` python context manager)
+//
+// Profiler setups a pair of callbacks that record profiling events and save them
+// into the thread local profiler struct (ThreadLocalDebugInfo, PROFILER slot)
+//
+//
+// Thus, the overall logic is:
+//
+// enableProfiler:
+//  - pushes new ThreadLocalDebugInfo (slot PROFILER) as the profiler config
+//    for the current thread
+//  - if needed, pushes profiling callbacks for the current thread
+//
+// disableProfiler:
+//  - pops PROFILER slot from the current ThreadLocalDebugInfo and
+//    consolidates events
+//  - if no one else is using profiler, pops profiling callbacks
+//
+// ThreadLocalState:
+//  - propagates ThreadLocalDebugInfo across threads
+//  - propagates profiler_callbacks_pushed_ TLS flag used to push
+//    profiling callbacks when they are needed and pop them
+//    when profiling is finished
+//
+// Profiler callbacks:
+//  - get the current profiling state (PROFILER slot in ThreadLocalDebugInfo)
+//  - saves profiling events into the profiling state
+//
+
+// TLS flag keeps track of (potentially nested) profiling ranges
+thread_local int64_t profiler_nested_depth_ = 0;
+
+// Profiler state
+struct ProfilerThreadLocalState : public at::DebugInfoBase {
```

Comment:
just a nit - we probably should rename DebugInfoBase into ThreadLocalStateBase or something. It's not about just debug anymore

Reply:
ThreadLocalState is a snapshot of settings while this is more a specific piece of data attached to the thread, but agree Debug is too narrow

Reply:
let's do this in a separate PR later, not to overcomplicate this one